### ISSUE TITLE
Unexpected behavior with range types and custom operators

### ISSUE THREAD

**Issue by fierce-goose** [nim-lang/Nim#24032]
### Description

```nim
proc `++`(n: var int) =
    n += 1

type
    r = range[ 0..15 ]

var a: r = 15

++a

echo a
```

this code should have thrown an error because I'm trying to assign a variable a value larger than possible, but it just goes beyond the allowed limits without error

### Nim Version

Nim Compiler Version 2.0.8 [Linux: amd64]

### Current Output

```text
16
```

### Expected Output

```text
sysFatal
Error: unhandled exception: over- or underflow [OverflowDefect]
```

**Comment by juancarlospaco** [comment]
!nim c

```nim
proc `++`(n: var int) = n += 1
type r = range[0..15]
var a: r = 15
++a
echo a
```

**Comment by beef331** [comment]
The issue is that `range` implicitly converted to `var int` as such it loses range checking. A work around is to use
```nim
proc `++`[T: range | SomeInteger](n: var T) =
    n = n + 1
  
type
    r = range[ 0..15 ]

var a: r = 15

++a

echo a
```

**Comment by juancarlospaco** [comment]
That should be in the documentation.

**Comment by Araq** [comment]
`range int` should not convert to `var int`.

**Comment by metagn** [comment]
Seems like a typo in the compiler

https://github.com/nim-lang/Nim/blob/aa0d8e9cfee7b15f80bd147b9bcf626012abb093/compiler/sigmatch.nim#L1020

`subtypeCheck` is only called for pointer types, but it also checks if the element type of the pointer type (`f.last`) is a pointer type which I don't think is intentional

I know changing this breaks a package though (from #23176) so we need to at least give a error message that suggests doing `++int(a)` instead ("explicit conversion needed" etc), but also we can only do this for `isSubrange`

**Comment by metagn** [comment]
Actually some subtype relations for `var` are allowed, like `openarray` and non-ref object subtypes:

```nim
type
  Foo = object of RootObj
  Bar = object of Foo

proc foo(x: var Foo) = echo "a"
var x: Bar
foo(x) # a
```

This is intentional: https://github.com/nim-lang/Nim/commit/f348671ba7146420e1bfc7cd21ddc5464f0b9cb3

But still `isIntConv` etc should fail. So I'm guessing only subtypes should be special cased to have this behavior.

Edit: Never mind the `openarray` relation is `isConvertible`. But still valid so just another special case.

Also `++int(a)` doesn't work

**Comment by Araq** [comment]
yes, it means "isSubrange" is a different type relation from "isSubtype". (I wasn't aware of this until now :-) )
